dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS/Node.pm

```diff
@@ -290,80 +364,1447 @@ sub new {
+# #if/#elsif/#else/#endif etc.
+#
+# Note that this only considers file-scoped C preprocessor directives;
+# ones within a code block such as CODE or BOOT don't contribute to the
+# state maintained here.
+#
+# Initially the whole XS part of the main XS file is considered a single
+# scope, so the single main cpp_scope node would have, as children, all
+# the file-scoped nodes such as Node::PROTOTYPES and any Node::xsub's.
+#
+# After an INCLUDE, the new XS file is considered as being in a separate
+# scope, and gets its own child cpp_scope node.
+#
+# Once an XS file starts having file-scope CPP conditionals, then each
+# branch of the conditional is considered a separate scope and  gets its
+# own cpp_scope node. Nested conditionals cause nested cpp_scope objects
+# in the AST.
+#
+# The main reason for this node type is to separate out the AST into
+# separate sections which can each have the same named XSUB without a
+# 'duplicate XSUB' warning, and where newXS()-type calls can be added to
+# to the boot code for *both* XSUBs, guarded by suitable #ifdef's.
+#
+# This node is the main high-level node where file-scoped parsing takes
+# place: its parse() method contains a fetch_para() loop which does all
+# the looking for file-scoped keywords, CPP directives, and XSUB
+# declarations. It implements a recursive-decent parser by creating child
+# cpp_scope nodes and recursing into that child's parse() method (which
+# does its own fetch_para() calls).
+
+BEGIN { $build_subclass->(
+    'type',       # Str:  what sort of scope: 'main', 'include' or 'if'
+    'is_cmd',     # Bool: for include type, it's INCLUDE_COMMAND
+    'guard_name', # Str:  the name of the XSubPPtmpAAAA guard define
+    'seen_xsubs', # Hash: the names of any XSUBs seen in this scope
+)};
+
+sub parse {
+    my __PACKAGE__        $self   = shift;
+    my ExtUtils::ParseXS  $pxs    = shift;
+
+    # Main loop: for each iteration, parse the next 'thing' in the current
+    # paragraph, such as a C preprocessor directive, a contiguous block of
+    # file-scoped keywords, or an XSUB. When the current paragraph runs
+    # out, read in another one. XSUBs, TYPEMAP and BOOT will consume
+    # all lines until the end of the current paragraph.
+    #
+    # C preprocessor conditionals such as #if may trigger recursive
+    # calls to process each branch until the matching #endif. These may
+    # cross paragraph boundaries.
+
+    while ( ($pxs->{line} && @{$pxs->{line}}) || $pxs->fetch_para())
+    {
+        next unless @{$pxs->{line}}; # fetch_para() can return zero lines
+
+        if (   !defined($self->{line_no})
+            && defined $pxs->{line_no}[0]
+        ) {
+            # set file/line_no after line number info is available:
+            # typically after the first call to fetch_para()
+            $self->SUPER::parse($pxs);
+        }
+
+        # skip blank line
+        shift @{$pxs->{line}}, next
+            if  @{$pxs->{line}} && $pxs->{line}[0] !~ /\S/;
```

Comment:
Do we need the ` @{$pxs->{line}}` condition check here?  This is already checked at the top of the loop.